doc/decisions/error_message.md

```diff
@@ -0,0 +1,133 @@
+# Error message & handling concept
+
+## Problem
+
+The current error concept has disadvantages in following regards:
+
+* Too verbose error message
+* A lot of redundant errors
+* Hard to manage specification file
+* No senseful way for application developers to use error codes from elektra
+
+This concept should improve useability for end users and developers.
+
+## Constraints
+
+* Error numbers must stay because they are more reliable to check against than strings
+
+## Assumptions
+
+## Considered Alternatives
+
+* Removing the specification file without requiring error numbers
+* Possible variations on what message should be displayed
+* Adding the key of the occurred error to the API which permits reading information from
+additional metadata such as an error message provided by a specification author. 
+Reason against: The description of the key should already provide such information.
+Doing it in an extra key would imply redundant information.
+
+## Decision
+
+The error message has the current format:
+
+```
+The command kdb set failed while accessing the key database with the info:
+Sorry, the error (#121) occurred ;(
+Description: validation failed
+Reason: Validation of key "<key>" with string "<value>" failed.
+Ingroup: plugin
+Module: enum
+At: ....../src/plugins/enum/enum.c:218
+Mountpoint: <parentKey>
+Configfile: ...../<file>.25676:1549919217.284067.tmp
+```
+
+The new default error message will look like this:
+```
+Sorry, plugin <PLUGIN> issued error #<NR>:
+Validation of key "<key>" with string "<value>" failed.
+```
+The error #<NR> will be the color red while <PLUGIN> will be the color blue.
+
+Optionally a third line indicating a solution can be added. Eg. for a permission related error there would be a third line:
+```
+Possible Solution: Retry the command as sudo (sudo !!)
+```
+
+To avoid losing information, the user can use the command line argument `-v` (verbose) to show
+`Mountpoint`, `Configfile` in addition to the current error message.
+Furthermore a developer can use the command line argument `-d` (debug) 
+to show `At` for debugging purposes.
+
+Warnings will be removed from the specification file as they are not needed there.
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
+
+Unused marked errors will be removed from the specification.
+
+The remaining ~130 errors will be categorizes into logical groups with subgroups.
+Each group and subgroup will receive a range of numbers such as in the HTTP protocol.
+
+* Permanent errors (1-500)
```

Comment:
Thank you for that input. Your idea also works nice. But if I understood correctly, theses error codes would then be strings because permanent errors would still have digits but something like 0100 for Resource errors. My thaught is that devs check for error codes like >=100 && <200 instead of something like charAt(0) == 0 && charAt(1) == 1 (or startsWith(01)).

Lets wait if @markus2330 has another idea/input on this :). For me, both approaches are equally fine.

Reply:
I would prefer if users do not have to fiddle around with numbers. @Piankero did you already did a study what other FLOSS projects are doing? You can write the comparison as thesis text.

In the next meeting we can do a brain storm of which projects have nice error concepts.

Reply:
Not yet, do you have a preference towards certain projects? Do they have to be similar to elektra (C/C++, Medium Large, Well Maintained, etc.)?

Reply:
I think we should look into projects that also have a generic plugin system where it is not so clear which kind of errors plugins might have to yield, e.g.: GStreamer, Apache http, Jenkins, Databases (MariaDB, PostgresQL).

Also relevant would be standards like SNMP.

Looking at projects that are to some extend alternatives to Elektra, like etcd, Windows Registry or Mac OS X plists also cannot harm.

Of course you should also look into POSIX errors (errno), although we definitely do not want to go back there as it is too primitive. (It does not give any reason or any further helpful information.)

As the error concepts are usually very short or non-existent, I think you can look at more than just a few.